rfcs/2023-07-03-theme-switching-ssr.md

```diff
@@ -0,0 +1,124 @@
+---
+Start Date: 03-07-2023
+RFC PR:
+Blade Issue:
+---
+
+# Theme Switching and Dark Mode with SSR / SSG
+
+## Summary
+
+This proposal lays out the approach for leveraging dynamic themes when using Blade along with Server Side Rendering (SSR) or Static Site Generation (SSG) without any [Flash of Unstyled Content (FoUC)](https://www.wikiwand.com/en/Flash_of_unstyled_content).
+
+> **Note**
+>
+> This proposal is only applicable for web apps (no changes for mobile apps).
+
+## Basic Example
+
+Consider the following example _(see steps below)_:
+
+![](./images/theme-switching-ssr/overview.png)
+
+Steps:
+
+1. User is on dark mode, browser sends request to server via CDN
+1. Server has no context of user preference _(see [alternatives](#alternatives))_
+1. Server serializes all the styles in default light mode _(we use CSS in JS in Blade)_ and sends back the HTML
+1. User receives the markup in light mode
+1. React [hydrates](https://react.dev/reference/react-dom/client/hydrateRoot) and changes the mode to match user preference (dark mode)
+1. There was a visible FoUC from the initial HTML rendered by browser till React hydrated
+
+## Motivation
+
+With Blade we can already do theme switching (eg. between `paymentLight` and `bankingDark`) using `ThemeProvider`. However, since we're using CSS in JS solution (`styled-components`), it means we can't do this theme switching before React has had a chance to hydrate.
+
+This leads to the primary motivation for this proposal:
+
+- With current approach components can't dynamically adjust to [user preferences](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) before hydrating / initial render
+- Current approach can lead to FoUC for products relying on SSR
+
+## Detailed Design
+
+> **Note**
+>
+> [Josh's blog](https://www.joshwcomeau.com/react/dark-mode) covers almost everything that is required and a recommended read before going through this RFC. In detailed design I'll cover things primarily from Blade's angle.
+
+### Rely on CSS variables for color related properties
+
+Currently, whenever we're applying color related properties on components (eg. `color`, `backgroundColor`) we read them off `theme` context. For example:
+
+```tsx
+// StyledAccordionButton.web.tsx
+// ...
+color: theme.colors.surface.action.icon.hover.lowContrast;
+// ...
+```
+
+Such occurrences lead to FoUC because `theme` is updated only after hydration. Now consider using [CSS variables](https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties) here:
+
+```tsx
+// StyledAccordionButton.web.tsx
+// ...
+color: var(--colors-surface-action-icon-hover-lowContrast);
```

Comment:
How would we switch the theme on client side? Any changes needed in BladeProvider?

Reply:
It should work like it's currently. Since we'll be picking color based tokens off css variables, we'll have to provide a setter utility for consumers to use for toggle functionality. Please check the linked blog.